src/arithmetic/canonical_simplify.cc

```diff
@@ -0,0 +1,811 @@
+      res = res % make_const(dtype, this->upper_factor);
+    }
+    if (this->lower_factor != 1) {
+      res = res / make_const(dtype, this->lower_factor);
+    }
+    sscale *= this->scale;
+    if (sscale != 1) {
+      CHECK(!dtype.is_uint() || sscale > 0);
+      res = res * make_const(dtype, sscale);
+    }
+    return res;
+  }
+
+  Expr Normalize() const final {
+    return NormalizeWithScale(1);
+  }
+
+  void MulToSelf(int64_t scale) {
+    this->scale *= scale;
+  }
+
+  inline bool IndexEqual(const SplitExpr& other) const;
+
+  /*! \brief positive infty */
+  static const constexpr int64_t kPosInf = ConstIntBoundNode::kPosInf;
+  static constexpr const char* _type_key = "arith.SplitExpr";
+  TVM_DECLARE_NODE_TYPE_INFO(SplitExprNode, CanonicalExprNode);
+};
+
+TVM_DEFINE_COW_NODE_REF(SplitExpr, Expr, SplitExprNode);
+
+inline bool SplitExprNode::IndexEqual(const SplitExpr& other) const {
+  if (index.same_as(other->index)) return true;
+  return ir::Equal(index, other->index);
+}
+
+/*!
+ * \brief Normal form that represent sum of expressions.
+ *
+ *  result = sum(args) + base.
+ */
+class SumExprNode : public CanonicalExprNode {
+ public:
+  /*!
+   * \brief arguments to be summed up.
+   *
+   * args are divided into segments with the same index.
+   * within each segment, the SplitExpr is ordered in descending order of lower_factor.
+   *
+   * \note Can be mutated by TryMergeSplitExpr, which is idempotent
+   */
+  mutable std::vector<SplitExpr> args;
+  /*! \brief Base value in the summation. */
+  int64_t base{0};
+
+  /*! \brief Simplify the args by merging SplitExprs */
+  void SimplifySplitExprs() const {
+    // NOTE: This algorithm relies on the factor that args are divided into segments
+    // and each segment is sorted in descending order of lower_factor.
+    for (size_t i = 0; i < args.size(); ++i) {
+      if (args[i]->scale == 0) continue;
+      for (size_t j = i + 1; j < args.size(); ++j) {
+        SplitExpr& lhs = args[i];
+        SplitExpr& rhs = args[j];
+        if (!lhs->IndexEqual(rhs)) break;
+        if (lhs->upper_factor < rhs->lower_factor) break;
+        if (lhs->lower_factor == rhs->upper_factor &&
+            lhs->scale % rhs->scale == 0 &&
+            lhs->lower_factor == (lhs->scale / rhs->scale) * rhs->lower_factor) {
+          // Merge condition:
+          // - x = lhs->index % lhs->upper_factor
+          // - s = lhs->scale / rhs->scale
+          // - c = rhs->lower_factor
+          //
+          //    ((x / (c * s)) * s + (x % (c * s)) / c
+          // => ((x / c) / s) * s + ((x / c) % s)
+          // => (x / c)
+          SplitExprNode* merged = rhs.CopyOnWrite();
+          merged->upper_factor = lhs->upper_factor;
+          // reset args[i] to be zero.
+          lhs.CopyOnWrite()->scale = 0;
+          break;
+        }
+      }
+    }
+    // sort by the entry
+    auto fcompare = [](const SplitExpr& lhs, const SplitExpr& rhs) {
+      // order by scale first
```

Comment:
Shouldn't it be ordered by `index` first? Or at least if the indices are different, the elements
should be incomparable.

Reply:
This is a good point, however, it is can be quite costly to deep compare indices. So instead we just order by the scale and factor so that it is mostly in a consistent form(ignoring the indices)

Reply:
Since the algorithm assumes that the vector contains contiguous segments of same-index elements, we have to check in the comparison function if lhs and rhs have the same index, otherwise this assumption may be destroyed by sorting. 

Also I would still suggest sorting by index because otherwise we may often get into the situation when something like `f(x + y) - f(y + x)` don't get simplified. And if it really leads to performance problems, we should think about optimizing deep comparison somehow (probably we can cache the size or some other measure of an expression).

Reply:
For now the code is fine because the result is only directly used by normalize, which is the intended usecase. It does break if we call it in the middle. 

The only reason why I am not sure about comparing index is that var comparison can depend on runtime, which makes its behavior indeterministic. I want to think a bit more about this before we come back to revisit it